File: ballast-core/src/dotnet/Models/Board.cs
using System;
using System.Collections.Generic;
using System.Linq;

namespace Ballast.Core.Models
{

    public class BoardState
    {
        public Guid Id { get; set; }
        public BoardType BoardType { get; set; }
        public TileShape TileShape { get; set; }
        public IEnumerable<Tile> Tiles { get; set; }
    }

    public class Board
    {

        public BoardType BoardType { get; private set; }
        public Guid Id { get; private set; }
        public IEnumerable<Tile> Tiles { get; private set; }
        public TileShape TileShape { get; private set; }

        private Board(BoardType boardType, Guid id, IEnumerable<Tile> tiles, TileShape tileShape)
        {
            BoardType = boardType;
            Id = id;
            Tiles = tiles;
            TileShape = tileShape;
        }
        
        public static Board FromProperties(BoardType boardType, Guid id, IEnumerable<Tile> tiles, TileShape tileShape) => new Board(
            boardType: boardType,
            id: id,
            tiles:tiles,
            tileShape: tileShape
        );

        public static implicit operator Board(BoardState state) =>
            new Board(state.BoardType, state.Id, state.Tiles, state.TileShape);

        public CubicCoordinates GetRandomPassableCoordinates()
        {
            var passables = Tiles.Where(x => x.Terrain.Passable);
            var index = new Random().Next(1, passables.Count()) - 1;
            return passables.ElementAt(index).CubicCoordinates;
        }

        public Tile GetTileFromCoordinates(CubicCoordinates cubicCoordinates) =>
            Tiles.SingleOrDefault(x => x.CubicCoordinates.Equals(cubicCoordinates));

    }
    
}

File: ballast-core/src/typescript/models/board.ts
import { BoardType, IBoardType } from './board-type';
import { CubicCoordinates, ICubicCoordinates } from './cubic-coordinates';
import { ITile, Tile } from './tile';
import { ITileShape, TileShape } from './tile-shape';

export interface IBoard {
    boardType: IBoardType;
    id: string;
    tiles: ITile[];
    tileShape: ITileShape;
}

export class Board implements IBoard {

    public readonly boardType: BoardType;
    public readonly id: string;
    public readonly tileShape: TileShape;
    public readonly tileMap: Map<string, Tile>;

    private cachedTiles?: Tile[];
    public get tiles() {
        if (!this.cachedTiles) {
            this.cachedTiles = Array.from(this.tileMap.values());
        }
        return this.cachedTiles;
    }

    public getTile(orderedTriple: number[]) {
        return this.tileMap.get(orderedTriple.toLocaleString());
    }

    private constructor(state: IBoard) {
        this.boardType = BoardType.fromObject(state.boardType);
        this.id = state.id;
        this.tileShape = TileShape.fromObject(state.tileShape);
        this.tileMap = new Map(this.mapTiles(state.tiles));
    }

    public static fromObject(object: IBoard) {
        return new Board(object);
    }

    private *mapTiles(tiles: ITile[]) {
        let tileIterator = Tile.fromObjectList(tiles);
        for (let tile of tileIterator) {
            let item: [string, Tile] = [ tile.cubicCoordinates.toOrderedTriple().toLocaleString(), tile ];
            yield item;
        }
    }

    public getRandomPassableCoordinates(): ICubicCoordinates {
        let passables = this.tiles
            .filter(x => !!x.terrain.passable);
        let index = Math.floor(Math.random() * Math.floor(passables.length)) - 1;
        return passables[index].cubicCoordinates;
    }

    public getTileFromCoordinates(cubicCoordinates: ICubicCoordinates) {
        return this.tiles.find(x => x.cubicCoordinates.equals(cubicCoordinates));
    }

}